### PR TITLE
Reverse autodiff reshape and slice

### PR DESCRIPTION
**Pull request by GeorgeNash**
Reverse mode auto diff for slice & reshape

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java [hunk 1]
```diff
@@ -26,8 +29,15 @@ protected DualNumber dualOp(DualNumber dualNumber) {
 
     @Override
     public Map<Vertex, PartialDerivatives> reverseModeAutoDifferentiation(PartialDerivatives derivativeOfOutputsWithRespectToSelf) {
-        //TODO
-        throw new UnsupportedOperationException("Reverse mode autodiff not supported for this operation");
+        Map<Vertex, PartialDerivatives> reshapedDerivatives = new HashMap<>();
+
+        for (Map.Entry<VertexId, DoubleTensor> partialDerivative : derivativeOfOutputsWithRespectToSelf.asMap().entrySet()) {
+            int[] newPartialShape = TensorShape.concat(getShape(), inputVertex.getShape());
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertex.java]: I think this only works when taking the derivative or this reshape wrt to something else.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/SliceVertexTest.java [hunk 2]
```diff
@@ -58,25 +63,59 @@ public void canRepeatablySliceForAPick() {
...
+    private void assertSlice(int dim, int ind, double[] expectedValue, int[] expectedShape, int[] expectedPartialShape) {
         DoubleVertex m = new UniformVertex(0, 10);
-        m.setValue(DoubleTensor.create(new double[]{1, 2, 3, 4}, 2, 2));
+        m.setValue(DoubleTensor.create(new double[]{1, 2, 3, 4, 6, 7, 8, 9}, 4, 2));
 
         DoubleVertex alpha = new UniformVertex(0, 10);
-        alpha.setValue(DoubleTensor.create(new double[]{10, 15, 20, 25}, 2, 2));
+        alpha.setValue(DoubleTensor.create(new double[]{10, 15, 20, 25, 30, 35, 40, 45}, 4, 2));
 
-        DoubleVertex N = m.matrixMultiply(alpha);
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/SliceVertexTest.java]: Changed it back to mmul, was just playing around with plus

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/SliceVertexTest.java [hunk 3]
```diff
@@ -58,25 +63,59 @@ public void canRepeatablySliceForAPick() {
...
+    public void sliceCorrectlySplitsRowOfPartialDerivativeDimOneIndexZero() {
+        assertSlice(1, 0, new double[]{11, 23, 36, 48}, new int[]{4, 1}, new int[]{4, 1, 4, 2});
+    }
+
+    @Test
+    public void sliceCorrectlySplitsRowOfPartialDerivativeDimOneIndexOne() {
+        assertSlice(1, 1, new double[]{17, 29, 42, 54}, new int[]{4, 1}, new int[]{4, 1, 4, 2});
+    }
+
+    private void assertSlice(int dim, int ind, double[] expectedValue, int[] expectedShape, int[] expectedPartialShape) {
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/binary/SliceVertexTest.java]: updated

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertexTest.java [hunk 4]
```diff
@@ -32,9 +37,17 @@ public void reshapeCorrectlyReshapesPartialDerivative() {
         DoubleVertex N = m.matrixMultiply(alpha);
 
         ReshapeVertex reshapedN = new ReshapeVertex(N, 4, 1);
-        DoubleTensor reshapedPartial = reshapedN.getDualNumber().getPartialDerivatives().withRespectTo(m);
 
-        Assert.assertArrayEquals(new int[]{4, 1, 2, 2}, reshapedPartial.getShape());
+        PartialDerivatives forward = reshapedN.getDualNumber().getPartialDerivatives();
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/ReshapeVertexTest.java]: yep - added two more

**Pull request merged**